File: typescript/src/receipt.ts
// Copyright (C) 2026, Lux Partners Limited. All rights reserved.
// See the file LICENSE for licensing terms.

import { keccak256, type Hex } from "viem";
import {
  DOMAIN_RECEIPT,
  RECEIPT_ENCODED_LEN,
  ReceiptStatus,
  address20,
  bytes32,
  bytesToHex,
  concatBytes,
  u16be,
  u256be,
  u64be,
  utf8,
} from "./bytesutil.js";

/**
 * InferenceReceipt is the cross-chain settlement receipt. Its canonical encoding
 * is the pinned 355-byte layout (encodeReceipt), byte-identical to
 * chains/aivm/receipts.go AInferenceReceipt and the Go SDK.
 */
export interface InferenceReceipt {
  version: number;
  intentId: Hex;
  taskId: Hex;
  cChainId: Hex;
  aChainId: Hex;
  requester: Hex;
  modelSpecHash: Hex;
  promptHash: Hex;
  canonicalOutputHash: Hex;
  status: ReceiptStatus;
  n: number;
  threshold: number;
  winnersRoot: Hex;
  operatorsRoot: Hex;
  feePaid: bigint;
  settledAtHeight: bigint | number;
}

/**
 * encodeReceipt produces the canonical 355-byte fixed-width encoding in the
 * pinned order:
 *
 *   u16be(version) || intentId(32) || taskId(32) || cChainId(32) || aChainId(32)
 *   || requester(20) || modelSpecHash(32) || promptHash(32) ||
 *   canonicalOutputHash(32) || u8(status) || u16be(n) || u16be(threshold) ||
 *   winnersRoot(32) || operatorsRoot(32) || u256be(feePaid,32) ||
 *   u64be(settledAtHeight)
 */
export function encodeReceipt(r: InferenceReceipt): Uint8Array {
  return concatBytes(
    u16be(r.version),
    bytes32(r.intentId),
    bytes32(r.taskId),
    bytes32(r.cChainId),
    bytes32(r.aChainId),
    address20(r.requester),
    bytes32(r.modelSpecHash),
    bytes32(r.promptHash),
    bytes32(r.canonicalOutputHash),
    Uint8Array.from([r.status & 0xff]),
    u16be(r.n),
    u16be(r.threshold),
    bytes32(r.winnersRoot),
    bytes32(r.operatorsRoot),
    u256be(r.feePaid),
    u64be(r.settledAtHeight),
  );
}

/** receiptHash = keccak256(DOMAIN_RECEIPT || encodeReceipt(r)). */
export function receiptHash(r: InferenceReceipt): Hex {
  return keccak256(bytesToHex(concatBytes(utf8(DOMAIN_RECEIPT), encodeReceipt(r))));
}

/** Only a Completed receipt with a non-zero output is actionable C-side. */
export function isCompleted(r: InferenceReceipt): boolean {
  return r.status === ReceiptStatus.Completed && !isZeroHash(r.canonicalOutputHash);
}

function isZeroHash(h: Hex): boolean {
  return /^0x0*$/.test(h);
}

/**
 * decodeReceipt parses a canonical 355-byte receipt encoding. Fail-secure:
 * rejects any input that is not exactly RECEIPT_ENCODED_LEN bytes.
 */
export function decodeReceipt(b: Uint8Array): InferenceReceipt {
  if (b.length !== RECEIPT_ENCODED_LEN) {
    throw new Error(`aichain: receipt length ${b.length}, want ${RECEIPT_ENCODED_LEN}`);
  }
  let o = 0;
  const rd = (n: number): Uint8Array => {
    const s = b.subarray(o, o + n);
    o += n;
    return s;
  };
  const hex = (n: number): Hex => bytesToHex(rd(n));
  const num = (n: number): bigint => {
    let v = 0n;
    for (const x of rd(n)) v = (v << 8n) | BigInt(x);
    return v;
  };
  const version = Number(num(2));
  const intentId = hex(32);
  const taskId = hex(32);
  const cChainId = hex(32);
  const aChainId = hex(32);
  const requester = hex(20);
  const modelSpecHash = hex(32);
  const promptHash = hex(32);
  const canonicalOutputHash = hex(32);
  const status = Number(num(1)) as ReceiptStatus;
  const n = Number(num(2));
  const threshold = Number(num(2));
  const winnersRoot = hex(32);
  const operatorsRoot = hex(32);
  const feePaid = num(32);
  const settledAtHeight = num(8);
  return {
    version,
    intentId,
    taskId,
    cChainId,
    aChainId,
    requester,
    modelSpecHash,
    promptHash,
    canonicalOutputHash,
    status,
    n,
    threshold,
    winnersRoot,
    operatorsRoot,
    feePaid,
    settledAtHeight,
  };
}

File: aichain/receipt.go
// Copyright (C) 2026, Lux Industries Inc. All rights reserved.
// See the file LICENSE for licensing terms.

package aichain

import (
	"fmt"
	"math/big"

	"github.com/luxfi/crypto"
	"github.com/luxfi/geth/common"
)

// InferenceReceipt is the cross-chain settlement receipt produced by the A-Chain
// when a Tier-2 inference task settles. Its canonical encoding is PINNED
// byte-for-byte (Encode / ReceiptEncodedLen = 355) and identical to
// chains/aivm/receipts.go AInferenceReceipt; its hash is
// keccak(DomainReceipt || Encode()), the leaf the receipt_root commits to.
//
// FeePaid is a non-negative *big.Int (nil == zero on the wire). Field order here
// is presentation; the WIRE order is fixed by Encode and MUST NOT be reordered.
type InferenceReceipt struct {
	Version             uint16         `json:"version"`
	IntentID            common.Hash    `json:"intentId"` // source C-chain intent id
	TaskID              common.Hash    `json:"taskId"`   // A-chain task id
	CChainID            common.Hash    `json:"cChainId"`
	AChainID            common.Hash    `json:"aChainId"`
	Requester           common.Address `json:"requester"`
	ModelSpecHash       common.Hash    `json:"modelSpecHash"`
	PromptHash          common.Hash    `json:"promptHash"`
	CanonicalOutputHash common.Hash    `json:"canonicalOutputHash"` // zero if Failed
	Status              uint8          `json:"status"`
	N                   uint16         `json:"n"`
	Threshold           uint16         `json:"threshold"`
	WinnersRoot         common.Hash    `json:"winnersRoot"`
	OperatorsRoot       common.Hash    `json:"operatorsRoot"`
	FeePaid             *big.Int       `json:"feePaid"`
	SettledAtHeight     uint64         `json:"settledAtHeight"`
}

// Encode produces the canonical fixed-width encoding (exactly ReceiptEncodedLen =
// 355 bytes) in the PINNED order, byte-identical to chains/aivm:
//
//	u16be(Version) || IntentID(32) || TaskID(32) || CChainID(32) || AChainID(32) ||
//	Requester(20) || ModelSpecHash(32) || PromptHash(32) || CanonicalOutputHash(32) ||
//	u8(Status) || u16be(N) || u16be(Threshold) || WinnersRoot(32) ||
//	OperatorsRoot(32) || u256be(FeePaid,32) || u64be(SettledAtHeight)
func (r InferenceReceipt) Encode() []byte {
	buf := make([]byte, 0, ReceiptEncodedLen)
	buf = append(buf, u16be(r.Version)...)
	buf = append(buf, r.IntentID.Bytes()...)
	buf = append(buf, r.TaskID.Bytes()...)
	buf = append(buf, r.CChainID.Bytes()...)
	buf = append(buf, r.AChainID.Bytes()...)
	buf = append(buf, r.Requester.Bytes()...)
	buf = append(buf, r.ModelSpecHash.Bytes()...)
	buf = append(buf, r.PromptHash.Bytes()...)
	buf = append(buf, r.CanonicalOutputHash.Bytes()...)
	buf = append(buf, r.Status)
	buf = append(buf, u16be(r.N)...)
	buf = append(buf, u16be(r.Threshold)...)
	buf = append(buf, r.WinnersRoot.Bytes()...)
	buf = append(buf, r.OperatorsRoot.Bytes()...)
	buf = append(buf, u256be(r.FeePaid)...)
	buf = append(buf, u64be(r.SettledAtHeight)...)
	return buf
}

// Hash is the receipt commitment: keccak256(DomainReceipt || Encode()). This is
// the leaf value (pre-leaf-hash) that goes into the receipt_root and that an
// exported proof proves membership of. Identical to chains/aivm
// AInferenceReceipt.Hash.
func (r InferenceReceipt) Hash() common.Hash {
	return common.BytesToHash(crypto.Keccak256(append([]byte(DomainReceipt), r.Encode()...)))
}

// Completed reports whether the receipt is the only actionable shape C-side:
// Status == StatusCompleted with a non-zero canonical output. A Pending,
// Failed, or Challenged receipt — or a zero output — is never credited.
func (r InferenceReceipt) Completed() bool {
	return r.Status == StatusCompleted && r.CanonicalOutputHash != (common.Hash{})
}

// DecodeReceipt parses a canonical 355-byte receipt encoding back into an
// InferenceReceipt. It is the inverse of Encode and rejects any input that is
// not EXACTLY ReceiptEncodedLen bytes (a corrupt/short/over-long frame is never
// reinterpreted into a credit — the same fail-secure discipline the precompile's
// DecodeReceipt applies).
func DecodeReceipt(b []byte) (InferenceReceipt, error) {
	var r InferenceReceipt
	if len(b) != ReceiptEncodedLen {
		return r, fmt.Errorf("aichain: receipt length %d, want %d", len(b), ReceiptEncodedLen)
	}
	o := 0
	rd16 := func() uint16 { v := uint16(b[o])<<8 | uint16(b[o+1]); o += 2; return v }
	rd32 := func() common.Hash { h := common.BytesToHash(b[o : o+32]); o += 32; return h }
	r.Version = rd16()
	r.IntentID = rd32()
	r.TaskID = rd32()
	r.CChainID = rd32()
	r.AChainID = rd32()
	r.Requester = common.BytesToAddress(b[o : o+20])
	o += 20
	r.ModelSpecHash = rd32()
	r.PromptHash = rd32()
	r.CanonicalOutputHash = rd32()
	r.Status = b[o]
	o++
	r.N = rd16()
	r.Threshold = rd16()
	r.WinnersRoot = rd32()
	r.OperatorsRoot = rd32()
	r.FeePaid = new(big.Int).SetBytes(b[o : o+32])
	o += 32
	var h uint64
	for i := 0; i < 8; i++ {
		h = h<<8 | uint64(b[o+i])
	}
	r.SettledAtHeight = h
	return r, nil
}
